libs/liblist-2.1/list.3
.TH LIST 3 "September 22, 1991"
.UC 6
.SH NAME
list_init, list_mvprev, list_mvnext, list_mvfront, list_mvrear, list_front, list_curr, list_rear, list_insert_before, list_insert_after, list_remove_front, list_remove_curr, list_remove_rear, list_size, list_empty, list_traverse, list_free \-- generic linked-list routines
.SH SYNOPSIS
.PP
cc [
.I flags
]
.IR files
.B -llist
[
.IR libraries
]
.nf
.PP
.ft B
\#include <list.h>
.PP
.ft B
LIST *list_init()
.PP
.ft B
LIST *list_mvprev(list)
    LIST *list;
.PP
.ft B
LIST *list_mvnext(list)
    LIST *list;
.PP
.ft B
LIST *list_mvfront(list)
    LIST *list;
.PP
.ft B
LIST *list_mvrear(list)
    LIST *list;
.PP
.ft B
char *list_front(list)
    LIST *list;
.PP
.ft B
char *list_curr(list)
    LIST *list;
.PP
.ft B
char *list_rear(list)
    LIST *list;
.PP
.ft B
char *list_insert_before(list, data, bytes)
    LIST *list;
    char *data;
    int bytes;
.PP
.ft B
char *list_insert_after(list, data, bytes)
    LIST *list;
    char *data;
    int bytes;
.PP
.ft B
char *list_remove_front(list)
    LIST *list;
.PP
.ft B
char *list_remove_curr(list)
    LIST *list;
.PP
.ft B
char *list_remove_rear(list)
    LIST *list;
.PP
.ft B
int list_size(list)
    LIST *list;
.PP
.ft B
int list_empty(list)
    LIST *list;
.PP
.ft B
int list_traverse(list, data, func, opts)
    LIST *list;
    char *data;
    int (*func)();
    int opts;
.PP
.ft B
void list_free(list, dealloc)
    LIST *list;
    void (*dealloc)();
.PP
.SH DESCRIPTION
These routines provide generic manipulation of (potentially)
multiple linked lists. Each list can hold arbitrarily sized elements, with
individual elements within a list varying in size. It is the
programmers responsibility to account for such differences.

Lists are referred to by variables declared as LIST *; the type LIST
is defined in <list.h> to be a C structure. Each descriptor maintains
references to the front, rear, and current elements, as well as the
size (or length) of the list. Various routines operate relative to the
current element.
.PP
.IR list_init
initializes and returns a list descriptor.
.PP
.IR list_mvprev ,
.IR list_mvnext ,
.IR list_mvfront ,
.IR list_mvrear ,
move to the previous, next, front, or rear element in
.IR list ,
and return the modified list descriptor. Movement previous or next is relative
to the current element.
.PP
.IR list_front ,
.IR list_curr ,
and
.IR list_rear
return pointers to the front, current, or rear element in
.IR list . \c



.PP
.IR list_insert_before
and
.IR list_insert_after
insert an element, pointed to by
.IR data
and of size
.IR bytes ,
into
.IR list ,
either before or after the current element. The newly inserted element is
then considered the current element. Both routines return a pointer to the
newly inserted data. If
.IR bytes
is greater than 0, then data is copied into the list, otherwise
only the reference
.IR data
is copied into the list. This allows the user to determine the
memory allocation policy.
.PP
.IR list_remove_front ,
.IR list_remove_curr ,
and
.IR list_remove_rear
remove the front, current, or rear element from
.IR list
and return a pointer to the removed data.
The current element is then set to the next element (prior to the remove) in
the list, unless the element at the rear of the list was removed. In this
case, the current element is set to the previous element (prior to the
remove).
.PP
.IR list_size
returns the size (or number of elements) of
.IR list
as an integer. \c



.PP
.IR list_empty
returns 1 (TRUE) if
.IR list
is empty, 0 (FALSE) otherwise. \c



.PP
.IR list_traverse
traverses
.IR list
according to
.IR opts ,
calling
.IR func
at each element encountered in the traversal, until
.IR func
returns 0 (FALSE) or the extent of the list is reached. This routine
can be used to search for an item or print the contents of the list,
for example. See the section LIST TRAVERSAL for more information.
.PP
.IR list_free
deallocates
.IR list,
applying the user-supplied function
.IR dealloc
to the data portion of each element remaining in the list. If
.IR dealloc
is LIST_DEALLOC, then the package will apply its own deallocation
routine. This, however, should only be done if the package has been
responsible for data element allocation, i.e., the insert routines
have been invoked with
.IR bytes
greater than 0. If
.IR dealloc
is LIST_NODEALLOC, no per-element deallocation will be performed.
.SH LIST TRAVERSAL
The behavior of the routine
.IR list_traversal
is controlled by the user-supplied function
.IR func
which is responsible for the actions performed at each element in the
traversal. In addition, the scope and direction of the
traversal can be specified with the
.IR opts
parameter.
.IR Func
should be declared as follows:
.PP
.RS
int func(data, curr)
    char *data;
    char *curr;
.RE
.PP
and should return 1 (TRUE) if the traversal should continue, or 0 (FALSE)
if the traversal should terminate. The function does not need to check
if the extent of the list has been reached; the return code from
.IR list_traverse
will indicate the status of the traversal. At each element encountered
in the traversal,
.IR list_traverse
will invoke
.IR func ,
passing it the two parameters
.IR data
and
.IR curr ;
.IR data
being the data pointer sent to
.IR list_traverse
originally, and
.IR curr
being the data pointer of the current element in the traversal. For example,
by passing the following function to
.IR list_traverse ,
the contents of a list of names could be printed.
.PP
.RS
.nf
int func(data, curr)
    char *data;
    char *curr;
{
    printf("Name=%30s.\\n", curr->name);
    return(TRUE);
}
.fi
.RE
.PP
In this example, the parameter
.IR data
is ignored and the function unconditionally returns 1 (TRUE), but functions
like
.PP
.RS
.nf
int func(data, curr)
    char *data;
    char *curr;
{
    if (strcmp(data, curr))
       return(TRUE);
    else
       return(FALSE);
}
.fi
.RE
.PP
can be used to position the current element pointer prior to insertion, so
as to keep the list ordered.
.PP
The direction and scope of the traversal can be controlled by specifying
one or more of the following options:
.PP
.RS
.nf
.ta 1.0i +1.0i +.5i
LIST_FORW * traverse forward (next)
LIST_BACK traverse backwards (prev)
LIST_FRNT * start from the front of the
   list (implies LIST_FORW)
LIST_CURR start from the current element
LIST_REAR start from the rear element
   (implies LIST_BACK)
LIST_SAVE * do not alter the current element
   pointer during the traversal
LIST_ALTR alter the current element pointer
   during the traversal
.fi
.RE
.PP
The asterisks (*) denote the default values. These options can be combined
with the logical OR operator, but at least one value must be specified.
For example, specifying
.PP
.RS
LIST_FORW
.RE
.PP
for
.IR opts
would request a traversal forwards from the current position,
restoring the current element pointer after the traversal, whereas
.PP
.RS
(LIST_BACK | LIST_CURR | LIST_ALTR)
.RE
.PP
would request a traversal backwards from the current position, and would
set the current element pointer to the last element encountered in the
traversal.

It should be noted that
.IR func
should not invoke any of the list routines unless LIST_ALTR has been
specified, since many of the routines act relative to the current
element pointer, which is not modified during a traversal with
LIST_SAVE specified.
.SH MEMORY ALLOCATION
The routines
.IR list_init ,
.IR list_insert_before ,
and
.IR list_insert_after
allocate memory during their execution. As such,
.IR list_insert_before
and
.IR list_insert_after
insert a copy of the data into the list when they are invoked with
.IR bytes
greater than 0. If
.IR bytes
is 0, then only the reference is copied into the list. This allows
the user to control the memory allocation policy. Both functions may
fail during memory allocation; see DIAGNOSTICS below for more information.

Note that
.IR list_remove_front ,
.IR list_remove_curr ,
and
.IR list_remove_rear
do not allocate memory for the removing data. They simply disassociate the
data from the list, and thus return a pointer to data that was previously
allocated by the package. It is the programmers responsibility to deallocate
such removed data.

If the user has been responsile for data element storage allocation, i.e. the
insert routines have been called with
.IR bytes
equal to 0, then the user must be responsible for storage deallocation
as well. A user-supplied deallocation function should be passed to
.IR list_free
for this purpose. The deallocation function should be declared as:
.PP
.RS
void dealloc(data)
   char *data;
.RE
.PP
and will be sent the data element reference of each element in the list
when
.IR list_free
is invoked. If the package has been responsible for data element allocation,
.IR list_free
can be invoked with LIST_DEALLOC for
.IR dealloc,;
the list package will apply its own deallocation routine, or
LIST_NODEALLOC if no per-element deallocation is required. It is the
programmers responsibility to insure that the memory allocation
policy is applied properly.
.SH DIAGNOSTICS
A NULL returned by
.IR list_init ,
.IR list_insert_before ,
or
.IR list_insert_after
indicates a failure in allocating memory for the new list or element. See
.IR malloc (3)
for more information.

.IR list_mvprev ,
.IR list_mvnext ,
.IR list_mvfront ,
.IR list_mvrear ,
.IR list_front ,
.IR list_curr ,
.IR list_rear ,
.IR list_remove_front,
.IR list_remove_curr,
and
.IR list_remove_rear
all return NULL if
.IR list
is empty.

.IR list_traverse
returns LIST_EMPTY for an empty list, LIST_EXTENT if an attempt was
made to move beyond the extent of the list, or LIST_OK otherwise.

A core dump indicates a bug ;-)
.SH BUGS
The routines
.IR list_remove_front ,
.IR list_remove_curr ,
.IR list_remove_rear ,
and
.IR list_free
do not physically reclaim storage space, although they do make it
available for reuse. While this is a function of
.IR free (3),
its application here could be considered a bug.
.SH SEE ALSO
queue(3), stack(3), cache(3)
.SH AUTHOR
Bradley C. Spatz (bcs@ufl.edu), University of Florida.


libs/liblist-2.1/list.c
/* list.c -- a generic list package
 * 
 * Last edited: Tue Jul 28 15:37:24 1992 by bcs (Bradley C. Spatz) on wasp
 *
 * Copyright (C) 1992, Bradley C. Spatz, bcs@ufl.edu
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 *
 * We define the following routines here:
 *
 *    LIST *list_init()
 *    LIST *list_mvprev(list)
 *    LIST *list_mvnext(list)
 *    char *list_insert_before(list, data, bytes)
 *    char *list_insert_after(list, data, bytes)
 *    char *list_remove_front(list)
 *    char *list_remove_curr(list)
 *    char *list_remove_rear(list)
 *    int list_traverse(list, data, func, opts);
 *    void list_free(list, dealloc)
 *
 * and the following here or in list.h if USE_MACROS defined for compilation.
 *
 *    char *list_front(list)
 *    char *list_curr(list)
 *    char *list_rear(list)
 *    LIST *list_mvfront(list)
 *    LIST *list_mvrear(list)
 *    int list_size(list)
 *    int list_empty(list)
 *
 * for
 *
 *    LIST *list;
 *    char *data;
 *    int bytes;
 *    int func(data, curr)
 *       char *data;
 *       char *curr;
 *    void dealloc(data)
 *       char *data;
 */

static char brag[] = "$$Version: list-2.1 Copyright (C) 1992 Bradley C. Spatz";

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>

#include "list.h"

LIST *list_init()
{
   LIST *list;

   /* Allocate, initialize, and return a new list. */
   list = (LIST *) malloc(sizeof(LIST));
   list->size = 0;
   list->front = NULL;;
   list->rear = NULL;;
   list->curr = NULL;;
   return(list);
}


LIST *list_mvprev(list)
LIST *list;
{
   /* Move to the previous link, if possible.  Note that the following
    * compound conditional expression *requires* a short-circuit evaluation.
    */
   if ((list->curr != NULL) && (list->curr->prev != NULL)) {
      list->curr = list->curr->prev;
      return(list);
   }
   else
      return(NULL);
}


LIST *list_mvnext(list)
LIST *list;
{
   /* Move to the next link, if possible.  Note that the following
    * compound conditional expression *requires* a short-circuit evaluation.
    */
   if ((list->curr != NULL) && (list->curr->next != NULL)) {
      list->curr = list->curr->next;
      return(list);
   }
   else
      return(NULL);
}


/* The following are definitions of these routines as functions.  We can
 * force thse to be implemented as macros by compiling with -DUSE_MACROS.
 * The macros are defined in the header file(s).  The macros afford better
 * performance, and if users know the routines are implemented as such,
 * they can always wrap their own functions aroud the macros if they need
 * function semantics (i.e. using the routines as pointers, as in passing
 * the routines as parameters to other functions.
 */
#ifndef USE_MACROS
LIST *list_mvfront(list)
LIST *list;
{
   /* Move to the front of the list.*/
   list->curr = list->front;
   return(list);
}


LIST *list_mvrear(list)
LIST *list;
{
   /* Move to the front of the list.*/
   list->curr = list->rear;
   return(list);
}


int list_empty(list)
LIST *list;
{
   /* Return 1 if the list is empty.  0 otherwise. */
   return((list->front == NULL) ? TRUE : FALSE);
}


char *list_front(list)
LIST *list;
{
   return((list->front == NULL) ? NULL : (list->front->data));
}


char *list_curr(list)
LIST *list;
{
   return((list->curr == NULL) ? NULL : (list->curr->data));
}


char *list_rear(list)
LIST *list;
{
   return((list->rear == NULL) ? NULL : (list->rear->data));
}


int list_size(list)
LIST *list;
{
   return(list->size);
}
#endif


static LIST_ELEMENT *list_create_element(data, bytes)
char *data;
int bytes;
{
   LIST_ELEMENT *new;

   /* Allocate storage for the new node and its data.  Return NULL if
    * unable to allocate.
    */
   new = (LIST_ELEMENT *) malloc(sizeof(LIST_ELEMENT));
   if (new == NULL) {
      return(NULL);
   }

   /* Allocate storage for the data only if requested; i.e. if bytes > 0.
    * Then either copy the data or just the reference into the node.
    */
   if (bytes > 0) {
      new->data = (char *) malloc(bytes);
      if (new->data == NULL) {
	 return(NULL);
      }
      (void) memcpy(new->data, data, bytes);
   }
   else {
      new->data = (char *) data;
   }

   return(new);
}


char *list_insert_before(list, data, bytes)
LIST *list;
char *data;
int bytes;
{
   LIST_ELEMENT *new;

   /* Allocate storage for the new element and its data.*/
   new = list_create_element(data, bytes);
   if (new == NULL)
      return(NULL);

   /* Now insert the element before the current, considering the cases:
    *    1) List is empty
    *    2) Inserting at front
    *    3) Otherwise
    * We handle them directly, in order.
    */
   if (list->front == NULL) {
      /* The list is empty.  Easy. */
      new->prev = new->next = NULL;
      list->front = list->rear = list->curr = new;
   }
   else if (list->curr->prev == NULL) {
      /* Inserting at the front. */
      new->prev = NULL;
      new->next = list->curr;
      list->curr->prev = new;
      list->front = new;
   }
   else {
      /* Otherwise. */
      new->prev = list->curr->prev;
      list->curr->prev->next = new;
      new->next = list->curr;
      list->curr->prev = new;
   }

   list->curr = new;
   list->size++;
   return(new->data);
}


char *list_insert_after(list, data, bytes)
LIST *list;
char *data;
int bytes;
{
   LIST_ELEMENT *new;

   /* Allocate storage for the new element and its data.*/
   new = list_create_element(data, bytes);
   if (new == NULL)
      return(NULL);

   /* Now insert the element after the current, considering the cases:
    *    1) List is empty
    *    2) Inserting at rear
    *    3) Otherwise
    * We handle them directly, in order.
    */
   if (list->front == NULL) {
      /* The list is empty.  Easy. */
      new->prev = new->next = NULL;
      list->front = list->rear = list->curr = new;
   }
   else if (list->curr->next == NULL) {
      /* Inserting at the rear. */
      new->next = NULL;
      new->prev = list->curr;
      list->curr->next = new;
      list->rear = new;
   }
   else {
      /* Otherwise. */
      new->next = list->curr->next;
      new->next->prev = new;
      new->prev = list->curr;
      list->curr->next = new;
   }

   list->curr = new;
   list->size++;
   return(new->data);
}


static char *list_remove_single(list)
LIST *list;
{
   char *data;

   /* The list has one element.  Easy. */
   data = list->curr->data;
   free(list->curr);
   list->front = list->rear = list->curr = NULL;
   list->size--;
   return (data);
}


char *list_remove_front(list)
LIST *list;
{
   LIST_ELEMENT *temp;
   char *data;

   /* Removing and return front element, or NULL if empty.  If curr
    * is the front, then curr becomes the next element.
    */
   if (list->front == NULL) {
      /* List is empty.  Easy. */
      return(NULL);
   }
   else if (list->front == list->rear) {
      /* List has only one element.  Easy. */
      data = list_remove_single(list);
   }
   else {
      /* List has more than one element.  Make sure to check if curr
       * points to the front.
       */
      data = list->front->data;
      list->front->next->prev = NULL;
      temp = list->front;
      list->front = temp->next;
      if (list->curr == temp)
	 list->curr = temp->next;
      free(temp);
      list->size--;
   }

   return(data);
}


char *list_remove_rear(list)
LIST *list;
{
   LIST_ELEMENT *temp;
   char *data;

   /* Removing and return rear element, or NULL if empty.  If curr
    * is the rear, then curr becomes the previous element.
    */
   if (list->front == NULL) {
      /* List is empty.  Easy. */
      return(NULL);
   }
   else if (list->front == list->rear) {
      /* List has only one element.  Easy. */
      data = list_remove_single(list);
   }
   else {
      /* List has more than one element.  Make sure to check if curr
       * points to the rear.
       */
      data = list->rear->data;
      list->rear->prev->next = NULL;
      temp = list->rear;
      list->rear = temp->prev;
      if (list->curr == temp)
	 list->curr = temp->prev;
      free(temp);
      list->size--;
   }

   return(data);
}


char *list_remove_curr(list)
LIST *list;
{
   LIST_ELEMENT *temp;
   char *data;

   /* Remove the current element, returning a pointer to the data, or
    * NULL if the list is empty.  Set curr to the next element unless
    * curr is at the rear, in which case curr becomes the previous
    * element.
    */
   if (list->front == NULL) {
      /* The list is empty.  Easy. */
      return(NULL);
   }
   else if (list->front == list->rear) {
      /* The list has one element.  Easy. */
      data = list_remove_single(list);
   }
   else if (list->curr == list->front) {
      /* Removing front element.  Easy. */
      data = list_remove_front(list);
   }
   else if (list->curr == list->rear) {
      /* Removing from the rear.  Easy.*/
      data = list_remove_rear(list);
   }
   else {
      /* Otherwise.  Must be inside a 3-element or larger list. */
      data = list->curr->data;
      temp = list->curr;
      temp->next->prev = temp->prev;
      temp->prev->next = temp->next;
      list->curr = temp->next;
      free(temp);
      list->size--;
   }

   return(data);
}


int list_traverse(list, data, func, opts)
LIST *list;
char *data;
int (*func)();
int opts;
{
   LIST_ELEMENT *lp;
   int status, rc;

   /* Traverse the list according to opts, calling func at each element,
    * until func returns 0 or the extent of the list is reached.  We return
    * 0 if the list is empty, 2 if we tried to go beyond the extent of the
    * list, and 1 otherwise.  We may or may not affect the current element
    * pointer.
    */
   if (list->front == NULL)
      return(LIST_EMPTY);
   
   /* Decide where to start. */
   if ((opts & LIST_CURR) == LIST_CURR) {
      lp = list->curr;
   }
   else if ((opts & LIST_REAR) == LIST_REAR) {
      lp = list->rear;
   }
   else {
      lp = list->front;
   }
   
   /* Now decide if to update the current element pointer. */
   if ((opts & LIST_ALTR) == LIST_ALTR)
      list->curr = lp;

   /* Now go until 0 is returned or we hit the extent of the list. */
   rc = LIST_OK;
   status = TRUE;
   while(status) {
      status = (*func)(data, lp->data);
      if (status) {
	 if ((((opts & LIST_BACK) == LIST_BACK) ? (lp->prev) : (lp->next))
	     == NULL) {
	    /* Tried to go beyond extent of list. */
	    status = FALSE;
	    rc = LIST_EXTENT;
	 }
	 else {
	    /* Decide where to go next. */
	    lp = (((opts & LIST_BACK) == LIST_BACK) ? (lp->prev) : (lp->next));

	    /* Now decide if to update the current element pointer. */
	    if ((opts & LIST_ALTR) == LIST_ALTR)
	       list->curr = lp;
	 }
      }
   }

   return(rc);
}


void list_free(list, dealloc)
LIST *list;
void (*dealloc)();
{
   char *data;

   /* Move to the front of the list.  Start removing elements from the
    * front.  Free up the data element by either applying the user-supplied
    * deallocation routine or free(3).  When we've gone through all the
    * elements, free the list descriptor.
    */
   list_mvfront(list);
   while (! list_empty(list)) {
      data = list_remove_front(list);
      /* Apply either no deallocation function to each node, our own, or
       * a user-supplied version.
       */
      if ((uintptr_t) dealloc != LIST_NODEALLOC) {
	 if ((uintptr_t) dealloc == LIST_DEALLOC) {
	    free(data);
	 }
	 else {
	    (*dealloc)(data);
	 }
      }
   }

   free(list);
}
